Sorting algos (#9)
Ada-Developers-Academy/core-unit-2: sorting-algos/selection-sort.md

# Selection Sort

## Learning Goals

- Describe selection sort and its efficiency

## Overview

The selection sort works by selecting the smallest unsorted item in the list and swapping it with index 0, then finding the next smallest and placing it into index 1 and so on.

### Detailed Explanation

![Selection Sort Example](../assets/sorting-algos/selection-sort.png)

Here is a more detailed explanation of the bubble sort algorithm:

1. You look through the entire array for the smallest element.
1. Once you find it, you swap this smallest element found with the first element of the array.
1. Then you look for the smallest element in the remaining array (the sub-array without the first element) and swap this element found with the second element.
1. Then you look for the smallest element in the remaining array (the sub-array without the first and second elements) and swap that element with the third element, and so on.

## Example

Consider the initial unsorted array `[99, 45, 35, 40, 16, 50, 11, 7, 90]`. Here's what it looks like before and after each iteration of the outer loop.

- `i` is the index that starts at 0, and increments until the length of the array.

> - `min_index` is a variable that holds the index of the smallest element in the sub-array.

Reviewer: since we're showing an abbreviated table  (unlike the fully exploded bubble sort), I'd like to expand the description for min_index to be very explicit that we need an inner loop to find the min value.

```suggestion
- `min_index` is a variable that holds the index of the smallest element remaining in the unsorted portion of the list. To locate this item, we use an inner loop through the remaining unsorted values for each pass of the outer loop.
```

> 

Reviewer: add guidance on how to interpret the following table

```suggestion

The array in each row is shown _before_ the swap for that pass is performed, so that the index under consideration `i` and the index of the minimum found value `min_index` line up with the displayed array values. The array shown on each following line has the previously identified swap applied.

```

> | Iteration | Array                                       | `i` | `min_index` |
> | --------- | ------------------------------------------- | --- | ----------- |
> | 1.        | [**99**, 45, 35, 40, 16, 50, 11, **7**, 90] | 0   | 7           |
> | 2.        | [7, **45**, 35, 40, 16, 50, **11**, 99, 90] | 1   | 6           |
> | 3.        | [7, 11, **35**, 40, **16**, 50, 45, 99, 90] | 2   | 4           |
> | 4.        | [7, 11, 16, **40**, **35**, 50, 45, 99, 90] | 3   | 4           |
> | 5.        | [7, 11, 16, 35, **40**, 50, 45, 99, 90]     | 4   | 4           |
> | 6.        | [7, 11, 16, 35, 40, **50**, **45**, 99, 90] | 5   | 6           |
> | 7.        | [7, 11, 16, 35, 40, 45, **50**, 99, 90]     | 6   | 6           |
> | 8.        | [7, 11, 16, 35, 40, 45, 50, **99**, **90**] | 7   | 8           |
> | 9.        | [7, 11, 16, 35, 40, 45, 50, 90, 99]         | -   |             |

Reviewer: similar table revisions as applied to bubble sort

```suggestion
| Pass | <div style="min-width: 300px;">Array</div>  | `i` | `min_index` |
| ---- | ------------------------------------------- | --- | ----------- |
| 1    | [**99**, 45, 35, 40, 16, 50, 11, **7**, 90] | 0   | 7           |
| 2    | [7, **45**, 35, 40, 16, 50, **11**, 99, 90] | 1   | 6           |
| 3    | [7, 11, **35**, 40, **16**, 50, 45, 99, 90] | 2   | 4           |
| 4    | [7, 11, 16, **40**, **35**, 50, 45, 99, 90] | 3   | 4           |
| 5    | [7, 11, 16, 35, **40**, 50, 45, 99, 90]     | 4   | 4           |
| 6    | [7, 11, 16, 35, 40, **50**, **45**, 99, 90] | 5   | 6           |
| 7    | [7, 11, 16, 35, 40, 45, **50**, 99, 90]     | 6   | 6           |
| 8    | [7, 11, 16, 35, 40, 45, 50, **99**, **90**] | 7   | 8           |
| 9    | [7, 11, 16, 35, 40, 45, 50, 90, 99]         | -   | -           |
```

> 

Reviewer: add observation about final pass

```suggestion

Note that we do not actually run the 9th pass for this 9 value array. Since there is only one item left in the unsorted portion of the array and no available swap locations, we can end one pass "early."

```

> ## Big(O) Complexity

Reviewer: ```suggestion
## Big O Complexity
```

> The time complexity of Selection sort is _O(n<sup>2</sup>)_.
> 
> Each of the nested loops has an upper bound defined by the count of elements, _n_ in the list to be sorted.

Reviewer: I think it's worth still spending some time repeating the big O discussion to build comfort with the topic. It also includes some verbiage acknowledging other languages and how concerns that affect one language may not be the same as those in another language.

```suggestion
The time complexity of selection sort is _O(n<sup>2</sup>)_.

The way by which we reach this conclusion is very similar to our discussion of bubble sort.

Our outer loop runs about _n_ times. The first pass through, we must perform _n-1_ comparisons to find the minimum value, and then perform a single swap for that value.

Note this minor improvement over bubble sort, as bubble sort would perform both _n-1_ comparisons, as well as _n-1_ swaps. This does not affect our overall complexity analysis since the inner loop does still run _n-1_ times. But again, in some languages this could be important.

Each successive pass requires one fewer comparison, so we end up with the same decreasing summation as we did in bubble sort, which gives us a complexity of _O(n<sup>2</sup>)_.

For a more intuitive argument. Consider that both the outer loop (tracking the position to place the minimum) and the inner loop (looking for the current minimum) are related to the size of the array _n_, and since they are nested, we multiply their bounds together for _n*n_ or _O(n<sup>2</sup>)_.
```

> 

Reviewer: include discussion of the sort stability

```suggestion

## Stability

Selection sort is _not_ a stable sorting algorithm. The swaps that we perform can result in equivalent values jumping over one another, losing the information about their relative initial orderings.

If we need a stable sorting algorithm, we should select a different algorithm, or think about how we could modify the base selection sort algorithm to enforce stability.

```

## Example Implementation

Consider this example implementation of selection sort.

```python
def selection_sort(array):
    i = 0
    while i < len(array) - 1:
        min_index = i
        j = i + 1
        while j < len(array):
            if array[j] < array[min_index]:
                min_index = j
            j += 1
        if min_index != i:
            temp = array[min_index]
            array[min_index] = array[i]
            array[i] = temp
        i += 1

>     return array

Reviewer: remove the return since this sorts in input array in-place

```suggestion
```

```

Compare this code with this detailed explanation of the algorithm:

> - Finds the smallest element in the range of `i` and `min_index`
> - Sets `min_index` to the index of the smallest item
> - If `i` and `min_index` are not the same, then
>   - Swap the elements at index `i` and index `min_index`
> - Increases `i`
> - Do this until `i` reaches the end of the list

Reviewer: - bring the explanation up to the same level of detail as the bubble sort
- the mention of min_index in the first line seems off to me anyway
- include final observation about the sort occurring to the passed in array directly.

```suggestion
a variable `i` at `0`
- Create an outer loop that loops while `i` is smaller than the length of the `array - 1` since we don't need to swap the final value
  - Start with the assumption that the minimum value is already in the right place by setting `min_index` to `i` 
  - Start a variable `j` at the position one past `i` to look for the smallest element in the rest of the array
  - Create an inner loop that loops `j` over the remainder of the array
    - If the current inspected value is smaller than the current smallest value, update `min_index` with the current index `j`
    - Increase `j`
  - Repeat until `j` reaches the end of the list
  - If `i` and `min_index` are not the same, then
    - Swap the elements at index `i` and index `min_index` using a temporary variable `temp`
  - Increase `i`
- Repeat until `i` reaches the end of the list

This implementation sorts the array in-place. Because the original array is modified, no return statement is needed.
```
